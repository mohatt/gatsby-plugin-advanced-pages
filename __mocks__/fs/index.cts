import { fs } from 'memfs'
import __realFs from 'fs'

/**
 * An internal reference to the real Node.js `fs` module.
 * Provides access to the native filesystem instance.
 */
Object.assign(fs, { __realFs })

module.exports = fs
