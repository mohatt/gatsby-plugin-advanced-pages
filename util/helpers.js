const path = require('path')
const debug = require("debug")("gatsby-plugin-advanced-pages")

// Checks if a path is a child of another one
function isPathChildof (child, parent) {
  const relative = path.relative(parent, child)
  return relative && !relative.startsWith('..') && !path.isAbsolute(relative)
}

module.exports = {
  debug,
  isPathChildof,
}
