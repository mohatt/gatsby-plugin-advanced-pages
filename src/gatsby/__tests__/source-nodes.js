import fs from 'fs'
import { sourceNodes } from '../'
import { mountFile, mountModule, mountOptions } from '../../../test/helpers'
import testCases from './__fixtures__/source-nodes'

// Use in-memory file system
jest.mock('fs')

describe('sourceNodes', () => {
  const createTypes = jest.fn()
  const createNode = jest.fn()
  const actions = { createTypes, createNode }
  const createNodeId = jest.fn().mockImplementation((v) => ('id-' + v).toUpperCase())
  const createContentDigest = jest.fn().mockImplementation((v) => JSON.stringify(v))
  const schema = { buildObjectType: jest.fn() }

  beforeEach(() => {
    createNode.mockReset()
    jest.resetModules()
    fs.reset()
    // Reset options before each test
    mountOptions()
  })

  for (const { title, files, throws, options } of testCases) {
    test(title, async () => {
      options && mountOptions(options)
      files.map(file => typeof file === 'string'
        ? mountFile(file)
        : Array.isArray(file.data)
          ? mountFile(file.path) || mountModule(file.path, file.data)
          : mountFile(file.path, file.data)
      )

      let error = null
      try {
        await sourceNodes({
          actions,
          schema,
          createNodeId,
          createContentDigest
        })
      } catch (e) {
        error = e
      }

      if (throws) {
        expect(error).not.toBeNull()
        expect(error).toMatchSnapshot()
        return
      }

      expect(error).toBeNull()
      expect(createNode.mock.calls).toMatchSnapshot()
    })
  }
})
