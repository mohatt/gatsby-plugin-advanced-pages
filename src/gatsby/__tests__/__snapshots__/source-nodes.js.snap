// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`sourceNodes correctly creates multiple page nodes with various configurations 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "data": Object {},
      "helper": "/virtual/project/gatsby/pages/helper.js",
      "id": "ID-0 - FOO",
      "internal": Object {
        "contentDigest": "{\\"id\\":\\"ID-0 - FOO\\",\\"title\\":\\"foo\\",\\"template\\":\\"/virtual/project/src/templates/default.js\\",\\"helper\\":\\"/virtual/project/gatsby/pages/helper.js\\",\\"routes\\":[{\\"name\\":\\"foo\\",\\"path\\":\\"/foo\\"},{\\"name\\":\\"foo.bar\\",\\"path\\":\\"/foo/bar\\"}],\\"data\\":{},\\"templateName\\":\\"default.js\\",\\"parent\\":null,\\"children\\":[],\\"internal\\":{\\"type\\":\\"Page\\",\\"description\\":\\"Advanced Page: foo\\"}}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": Array [
        Object {
          "name": "foo",
          "path": "/foo",
        },
        Object {
          "name": "foo.bar",
          "path": "/foo/bar",
        },
      ],
      "template": "/virtual/project/src/templates/default.js",
      "templateName": "default.js",
      "title": "foo",
    },
  ],
  Array [
    Object {
      "children": Array [],
      "data": Object {},
      "id": "ID-1 - LOREM",
      "internal": Object {
        "contentDigest": "{\\"id\\":\\"ID-1 - LOREM\\",\\"title\\":\\"lorem\\",\\"template\\":\\"/virtual/project/src/templates/test.js\\",\\"routes\\":[{\\"name\\":\\"lorem\\",\\"path\\":\\"/lorem\\"}],\\"data\\":{},\\"templateName\\":\\"test.js\\",\\"parent\\":null,\\"children\\":[],\\"internal\\":{\\"type\\":\\"Page\\",\\"description\\":\\"Advanced Page: lorem\\"}}",
        "description": "Advanced Page: lorem",
        "type": "Page",
      },
      "parent": null,
      "routes": Array [
        Object {
          "name": "lorem",
          "path": "/lorem",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "test.js",
      "title": "lorem",
    },
  ],
  Array [
    Object {
      "children": Array [],
      "data": Object {
        "name": "john",
        "title": "don",
      },
      "helper": "/virtual/project/gatsby/custom/helper.js",
      "id": "ID-2 - IPSUM",
      "internal": Object {
        "contentDigest": "{\\"id\\":\\"ID-2 - IPSUM\\",\\"title\\":\\"ipsum\\",\\"template\\":\\"/virtual/project/src/custom/template.js\\",\\"helper\\":\\"/virtual/project/gatsby/custom/helper.js\\",\\"data\\":{\\"name\\":\\"john\\",\\"title\\":\\"don\\"},\\"routes\\":[{\\"name\\":\\"ipsum\\",\\"path\\":\\"/ipsum\\"},{\\"name\\":\\"dolor\\",\\"path\\":\\"/dolor\\"},{\\"name\\":\\"sit\\",\\"path\\":\\"/sit/amet\\"}],\\"templateName\\":\\"./src/custom/template.js\\",\\"parent\\":null,\\"children\\":[],\\"internal\\":{\\"type\\":\\"Page\\",\\"description\\":\\"Advanced Page: ipsum\\"}}",
        "description": "Advanced Page: ipsum",
        "type": "Page",
      },
      "parent": null,
      "routes": Array [
        Object {
          "name": "ipsum",
          "path": "/ipsum",
        },
        Object {
          "name": "dolor",
          "path": "/dolor",
        },
        Object {
          "name": "sit",
          "path": "/sit/amet",
        },
      ],
      "template": "/virtual/project/src/custom/template.js",
      "templateName": "./src/custom/template.js",
      "title": "ipsum",
    },
  ],
]
`;

exports[`sourceNodes correctly creates page nodes from pages.config.js 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "data": Object {},
      "id": "ID-0 - FOO",
      "internal": Object {
        "contentDigest": "{\\"id\\":\\"ID-0 - FOO\\",\\"title\\":\\"foo\\",\\"template\\":\\"/virtual/project/src/templates/test.js\\",\\"routes\\":[{\\"name\\":\\"foo\\",\\"path\\":\\"/foo\\"}],\\"data\\":{},\\"templateName\\":\\"./src/templates/test.js\\",\\"parent\\":null,\\"children\\":[],\\"internal\\":{\\"type\\":\\"Page\\",\\"description\\":\\"Advanced Page: foo\\"}}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": Array [
        Object {
          "name": "foo",
          "path": "/foo",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo",
    },
  ],
]
`;

exports[`sourceNodes correctly creates page nodes from pages.config.json 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "data": Object {},
      "id": "ID-0 - FOO",
      "internal": Object {
        "contentDigest": "{\\"id\\":\\"ID-0 - FOO\\",\\"title\\":\\"foo\\",\\"template\\":\\"/virtual/project/src/templates/test.js\\",\\"routes\\":[{\\"name\\":\\"foo\\",\\"path\\":\\"/foo\\"}],\\"data\\":{},\\"templateName\\":\\"./src/templates/test.js\\",\\"parent\\":null,\\"children\\":[],\\"internal\\":{\\"type\\":\\"Page\\",\\"description\\":\\"Advanced Page: foo\\"}}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": Array [
        Object {
          "name": "foo",
          "path": "/foo",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo",
    },
  ],
]
`;

exports[`sourceNodes correctly creates page nodes from pages.config.yaml 1`] = `
Array [
  Array [
    Object {
      "children": Array [],
      "data": Object {},
      "id": "ID-0 - FOO",
      "internal": Object {
        "contentDigest": "{\\"id\\":\\"ID-0 - FOO\\",\\"title\\":\\"foo\\",\\"template\\":\\"/virtual/project/src/templates/test.js\\",\\"routes\\":[{\\"name\\":\\"foo\\",\\"path\\":\\"/foo\\"}],\\"data\\":{},\\"templateName\\":\\"./src/templates/test.js\\",\\"parent\\":null,\\"children\\":[],\\"internal\\":{\\"type\\":\\"Page\\",\\"description\\":\\"Advanced Page: foo\\"}}",
        "description": "Advanced Page: foo",
        "type": "Page",
      },
      "parent": null,
      "routes": Array [
        Object {
          "name": "foo",
          "path": "/foo",
        },
      ],
      "template": "/virtual/project/src/templates/test.js",
      "templateName": "./src/templates/test.js",
      "title": "foo",
    },
  ],
]
`;

exports[`sourceNodes throws and error for invalid pages config file 1`] = `
[Error: Unable to validate pages config file "/virtual/project/pages.config.yaml":
 "[0].routes.foo": Route path must be a string starting with '/']
`;

exports[`sourceNodes throws and error for non existing pages config file 1`] = `
[Error: "gatsby-plugin-advanced-pages" might not be working properly:
 Unable to find a valid pages config file under "/virtual/project"]
`;

exports[`sourceNodes throws and error for pages with invalid template 1`] = `
[Error: A path with value "./no/template.js" could not be found at any of the following locations:
 - "/virtual/project/src/templates/no/template.js
 - "/virtual/project/no/template.js"]
`;

exports[`sourceNodes throws and error when pages don not have a valid template 1`] = `[Error: Missing "template" metadata for page[0]: "foo". No default template is set in plugin options either.]`;
